google-cloud-spanner/src/main/java/com/google/cloud/spanner/Mutation.java

```diff
@@ -402,20 +404,43 @@ private boolean isFloat32NaN(Value value) {
     return value.getType().equals(Type.float32()) && Float.isNaN(value.getFloat32());
   }
 
-  static void toProto(Iterable<Mutation> mutations, List<com.google.spanner.v1.Mutation> out) {
+  /**
+   * Converts the list of mutations to the corresponding protobuf mutations and returns a random
+   * mutation from the available list based on the following heuristics:
+   *
+   * <p>1. Prefer mutations other than INSERT, as INSERT mutations may contain autogenerated columns
+   * whose information is unavailable on the client.
+   *
+   * <p>2. If the list only contains INSERT mutations, select the one with the highest number of
+   * values.
+   */
+  static com.google.spanner.v1.Mutation toProtoAndReturnRandomMutation(
+      Iterable<Mutation> mutations, List<com.google.spanner.v1.Mutation> out) {
     Mutation last = null;
     // The mutation currently being built.
     com.google.spanner.v1.Mutation.Builder proto = null;
     // The "write" (!= DELETE) or "keySet" (==DELETE) for the last mutation encoded, for coalescing.
     com.google.spanner.v1.Mutation.Write.Builder write = null;
     com.google.spanner.v1.KeySet.Builder keySet = null;
+
+    // Stores all the mutations excluding INSERT mutations.
+    List<com.google.spanner.v1.Mutation> allMutationsExcludingInsert = new ArrayList<>();
+    // Stores the INSERT mutation with largest number of values.
+    com.google.spanner.v1.Mutation largestInsertMutation =
+        com.google.spanner.v1.Mutation.getDefaultInstance();
+
     for (Mutation mutation : mutations) {
       if (mutation.operation == Op.DELETE) {
         if (last != null && last.operation == Op.DELETE && mutation.table.equals(last.table)) {
           mutation.keySet.appendToProto(keySet);
         } else {
           if (proto != null) {
-            out.add(proto.build());
+            com.google.spanner.v1.Mutation builtMutation = proto.build();
+            out.add(builtMutation);
+            if (checkIfInsertMutationWithLargeValue(builtMutation, largestInsertMutation)) {
+              largestInsertMutation = builtMutation;
+            }
+            addMutationsExcludingInsert(builtMutation, allMutationsExcludingInsert);
```

Comment:
nit: name this method `maybeAddMutationToListExludingInserts(...)` (or something similar). The current naming seems to indicate that it is added unconditionally.

Reply:
Done